### PR TITLE
enhancement(splunk_hec sink): Implement indexer acknowledgements

### LINKED ISSUE #2374: Support indexer acknowledgements with `splunk_hec` sink
Apparently, when data is sent to a Splunk HEC it is not guaranteed to be saved if a 200 is received. To solve this Splunk implements an [indexer acknowledgment feature](https://docs.splunk.com/Documentation/Splunk/8.0.3/Data/AboutHECIDXAck). We should look into supporting within the `splunk_hec` sink.

### PR DESCRIPTION
**Pull request by 001wwang**
Closes #2374 

This PR
- Adds indexer acknowledgement support to the `splunk_hec` logs and metrics sink
- Updates the Splunk integration test script to use a Splunk HEC image with indexer acknowledgements enabled.
- Refactors the struct used to build http requests for `splunk_hec` sinks

Differences from RFC
- Dropped the [shared structure](https://github.com/vectordotdev/vector/blob/master/rfcs/2021-10-29-9465-splunk-hec-indexer-acknowledgements.md#implementation-1) in favor of using channels to send data back-and-forth to the ack query task.

### CODE REVIEW

File: src/sinks/splunk_hec/common/acknowledgements.rs
Comment (001wwang) [inline src/sinks/splunk_hec/common/acknowledgements.rs:32]: Note, there is some overlap in necessary structures with the [source-side implementation](https://github.com/vectordotdev/vector/pull/10044/files#diff-d4466a02a0d126c29c9f5f53abce0f98456dbc58c423210d4b80500dbd67f87cR214). Will consolidate these once the source-side implementation is merged.

Comment (001wwang) [reply src/sinks/splunk_hec/common/acknowledgements.rs:32]: On second thought, the duplication is minor, and it seems better/simpler not to couple the source and sink in that way. This leaves room to adjust the sink-side implementation if needed to support future Splunk HEC versions while keeping our source-side implementation fixed. I'll leave it as is for now.

**Review by blt — APPROVED** [review]

File: src/sinks/splunk_hec/common/acknowledgements.rs
Comment (bruceg) [inline src/sinks/splunk_hec/common/acknowledgements.rs:197]: Could this ack client run in parallel with this receive loop to avoid the need for a select, and also avoid issues with delays fetching the acks from delaying receiving new data from the channel?

File: website/cue/reference/components/sinks/splunk_hec.cue
Comment (binarylogic) [inline website/cue/reference/components/sinks/splunk_hec.cue:5]: To clarify, this deviates from other sink's `acknowledgements` configuration style (`acknowledgements.enabled`) because it's automatic, correct? It turns on when response includes an `ackID`. Do we not want to let users opt-out if they don't want this, for performance reasons, etc?

File: website/cue/reference/components/sinks/splunk_hec.cue
Comment (binarylogic) [inline website/cue/reference/components/sinks/splunk_hec.cue:33]: I'd like to see us address how this works with Vector's internal acknowledgements feature. Does it tie in appropriately? For example, will ack's flow back up the pipeline to the source?

File: src/sinks/splunk_hec/common/service.rs
Comment (bruceg) [inline src/sinks/splunk_hec/common/service.rs:34]: Does this need to be global to the entirety of Vector, or could it be a data member in `HecService`? Also, since it's a fixed-length string, could this be the same `[u8; 45]` as the `encode_buffer` returns?

File: website/cue/reference/components/sinks/splunk_hec.cue
Comment (binarylogic) [inline website/cue/reference/components/sinks/splunk_hec.cue:38]: We use the [`urls.cue`](https://github.com/vectordotdev/vector/blob/ab0744ae8ea1557166257fd1308fac585f751c60/website/cue/reference/urls.cue) file for this. It helps to ensure we have unique links, validate that they work, etc.
```suggestion
				[Splunk HEC indexer acknowledgements](\(urls.splunk_hec_indexer_acknowledgements))
```

File: src/sinks/splunk_hec/common/service.rs
Comment (bruceg) [inline src/sinks/splunk_hec/common/service.rs:46]: Do you actually use this field (except in test code)? If not, you can just drop it from the struct as you aren't using `deny_unknown_fields` and skip the expense of handling the string. In fact, it doesn't look like you are using the `code` either. If you need it for tests, you can redefine the struct there.

File: src/sinks/splunk_hec/common/service.rs
Comment (bruceg) [inline src/sinks/splunk_hec/common/service.rs:374]: Should this be `Errored`, as the server didn't explicitly reject the events?

File: src/sinks/splunk_hec/common/service.rs
Comment (bruceg) [inline src/sinks/splunk_hec/common/service.rs:339]: Am I right in thinking this is the only way the server can signal rejection? If there is another way to signal rejection, then this might need another test.

**Review by bruceg — COMMENTED** [review]
It looks like a lock file `website/.hugo_build.lock` leaked into this PR

File: website/cue/reference/components/sinks/splunk_hec.cue
Comment (bruceg) [reply website/cue/reference/components/sinks/splunk_hec.cue:5]: Sinks do not normally have acknowledgement configuration. If the protocol allows, they always wait for and check the response from a request.

Comment (001wwang) [reply website/cue/reference/components/sinks/splunk_hec.cue:5]: Right, the `splunk_hec` sinks either acknowledge based on an initial HTTP `200 OK` response status or based on this indexer acknowledgements implementation (if an `ackID` is present). I support making the indexer acknowledgements behavior explicitly disable-able since it can add potentially significant overhead to the source. Instead of `acknowledgements.enabled`, maybe this can look like `acknowledgements.indexer_acknowledgements_enabled`.

A current workaround to opt-out would be for users to configure the sink with a Splunk HEC token that has indexer acknowledgements disabled.

File: website/cue/reference/components/sinks/splunk_hec.cue
Comment (001wwang) [reply website/cue/reference/components/sinks/splunk_hec.cue:33]: Sure, I can flesh this out. It does tie in to the end-to-end acknowledgements feature and makes it more accurate when used with `splunk_hec` sinks. The current behavior of acking solely based on a `200 OK` response is not as robust since Splunk HEC can still lose the data after that response.

File: src/sinks/splunk_hec/logs/integration_tests.rs
Comment (001wwang) [reply src/sinks/splunk_hec/logs/integration_tests.rs:341]: Do you mean a scenario like:
- Send events to Splunk and receive `ack_id`'s
- `ack_id`'s are never acked and expire into `EventStatus::Failed`

I don't think there's a way to force that as that acking failure would be up to Splunk internals.

File: src/sinks/splunk_hec/metrics/config.rs
Comment (001wwang) [reply src/sinks/splunk_hec/metrics/config.rs:25]: No, I generally stick to auto-import via VSCode and whatever formatting `make fmt` provides. Inconsistencies may be from accidental modifications or rare moments when I do things manually -- I'll clean this up.

File: src/sinks/splunk_hec/common/acknowledgements.rs
Comment (001wwang) [reply src/sinks/splunk_hec/common/acknowledgements.rs:197]: If they ran in parallel, I think we'd need a lock around `HecAckClient.acks`. Running the ack client may require modifying `acks` and receiving new `ack_id`'s also modifies `acks`. Would the required locking offset the benefits of parallelizing?

File: src/sinks/splunk_hec/common/service.rs
Comment (001wwang) [reply src/sinks/splunk_hec/common/service.rs:34]: I'll refactor creating this logic into `HecService`. I forgot my original intentions with this method -- the purpose may have been lost after some refactoring.

File: src/sinks/splunk_hec/common/service.rs
Comment (001wwang) [reply src/sinks/splunk_hec/common/service.rs:46]: Gotcha, will do. The fields are not currently used.

File: src/sinks/splunk_hec/common/service.rs
Comment (001wwang) [reply src/sinks/splunk_hec/common/service.rs:339]: Right, currently the only way to receive `EventStatus::Failed` is for all of an `ack_id`'s retries to expire. In that case, the `tx` used to send back an `EventStatus` is dropped and we default to `Failed`.

File: src/sinks/splunk_hec/common/service.rs
Comment (001wwang) [reply src/sinks/splunk_hec/common/service.rs:374]: I see. I've had a hard time with the error handling, so it's somewhat messy. In this case, an `ack_id` would have been queried `retry_limit` number of times and never been successfully acked. It would then be dropped and result in `EventStatus::Failed` since we don't keep track of whether it hasn't been acked due to `ack_id: false` status or because the endpoint is erroring. 

Supporting `EventStatus::Errored` seems more accurate but I'm curious if you have suggestions for the implementation.

File: website/cue/reference/components/sinks/splunk_hec.cue
Comment (001wwang) [reply website/cue/reference/components/sinks/splunk_hec.cue:33]: I added this detail in 6938ae2, let me know if that makes sense.

Comment (binarylogic) [reply website/cue/reference/components/sinks/splunk_hec.cue:33]: Looks good to me.

File: src/sinks/splunk_hec/common/service.rs
Comment (bruceg) [reply src/sinks/splunk_hec/common/service.rs:367]: I think `Errored` is more accurate as well as we don't know why the events were never acked.

File: src/sinks/splunk_hec/common/acknowledgements.rs
Comment (bruceg) [reply src/sinks/splunk_hec/common/acknowledgements.rs:194]: Yes, we would either need a lock or similar. My concern was if there was a delay in the process of fetching the acks, the new ids could pile up waiting behind that process. This doesn't have to be blocking if others think this is a minor concern.

File: src/sinks/splunk_hec/logs/integration_tests.rs
Comment (bruceg) [reply src/sinks/splunk_hec/logs/integration_tests.rs:341]: I meant some kind of explicit negative acknowledgement, but I later saw this is not possible. I think the other tests cover that.

File: src/sinks/splunk_hec/common/service.rs
Comment (001wwang) [reply src/sinks/splunk_hec/common/service.rs]: Updated in e81824d

File: src/sinks/splunk_hec/common/acknowledgements.rs
Comment (001wwang) [reply src/sinks/splunk_hec/common/acknowledgements.rs:206]: Once merged, I'll add a soak test with `splunk_hec source` -> possible transform -> `splunk_hec sink` (the lading splunk generator/blackhole have acking behavior). I'm hoping that will show if there are significant issues.

File: src/sinks/splunk_hec/common/acknowledgements.rs
Comment (jszwedko) [inline src/sinks/splunk_hec/common/acknowledgements.rs:80]: Usually we like to separate out fields, even for debug messages. This would be something like:

```rust
debug!(message = "Received ack statuses", ?ack_query_response);
```

File: src/sinks/splunk_hec/common/acknowledgements.rs
Comment (jszwedko) [inline src/sinks/splunk_hec/common/acknowledgements.rs:97]: Should this be an internal event with a metric?

File: src/sinks/splunk_hec/common/acknowledgements.rs
Comment (jszwedko) [inline src/sinks/splunk_hec/common/acknowledgements.rs:105]: Same here, internal event?

File: src/sinks/splunk_hec/common/acknowledgements.rs
Comment (jszwedko) [inline src/sinks/splunk_hec/common/acknowledgements.rs:211]:
```suggestion
                        debug!(message = "Stored ack id", ?ack_id);
```

Same feedback as above about separating the message and fields.

File: src/sinks/splunk_hec/common/acknowledgements.rs
Comment (jszwedko) [inline src/sinks/splunk_hec/common/acknowledgements.rs:119]:
```suggestion
                debug!(message = "Finalized ack id", ?ack_id);
```

File: src/sinks/splunk_hec/common/service.rs
Comment (jszwedko) [inline src/sinks/splunk_hec/common/service.rs:102]: Should we log when this happens? Maybe another internal event?

**Review by jszwedko — COMMENTED** [review]
Nice work. The UX looks good to me: enabling if the downstream supports them as well as the exposed config options.

I left some nits about log messages and internal events.

I am wondering if we want to add some instrumentation here, though? As an operator, I'd probably be interested in the number of outstanding acks in the sink to know how my pipeline is performing. If my Splunk cluster is ack'ing slowly, I may need to investigate.

Similarly, should we have a cap on outstanding acks? Do you have an idea of the overhead of each one? If the down stream is acking too slowly, I'd prefer that Vector start applying back-pressure to the source than run out of memory.

File: src/sinks/splunk_hec/common/acknowledgements.rs
Comment (001wwang) [reply src/sinks/splunk_hec/common/acknowledgements.rs:97]: Sure thing, I'll refactor this and others mentioned to be more consistent with standard internal instrumentation.

**Comment by 001wwang** [comment]
> Also, could we expose an option to disable acks, even if the server supports them? I realize it is unlikely that users would want that, but I like the idea of being able to preserve the current behavior in-case we release and uncover a bug with this new behavior: to give users a way to fall-back until we can address.

Yes we could. This was discussed [a bit in this thread](https://github.com/vectordotdev/vector/pull/10135#discussion_r756452510), and the current opt-out would be to use a Splunk token with indexer acknowledgements disabled. I can see that being very cumbersome for existing setups, so I'm happy to add an explicit option to the configuration. Does an `acknowledgements.indexer_acknowledgements_enabled` option make sense?

File: src/sinks/splunk_hec/common/service.rs
Comment (001wwang) [reply src/sinks/splunk_hec/common/service.rs:102]: Sure, it could be useful for getting insight if this edge case is encountered.

**Comment by jszwedko** [comment]
👍 `acknowledgements.indexer_acknowledgements_enabled` makes sense to me.

**Comment by 001wwang** [comment]
> Similarly, should we have a cap on outstanding acks? Do you have an idea of the overhead of each one? If the down stream is acking too slowly, I'd prefer that Vector start applying back-pressure to the source than run out of memory.

For overhead, each ack requires a `u64` (for the id) and a `oneshot::channel` (for communicating status). I'm not sure what a typical number of outstanding acks would be, but I see the value in having a cap. From the Splunk side, the default number of ack statuses they store per channel is 1_000_000 which implies that, by default, a client would at most query for 1_000_000 acks. Is that a reasonable cap? Should we let this be user-configurable?

File: src/sinks/splunk_hec/common/service.rs
Comment (001wwang) [reply src/sinks/splunk_hec/common/service.rs]: Added an event with a `warning` log: 8d40e3d

**Comment by jszwedko** [comment]
> > Similarly, should we have a cap on outstanding acks? Do you have an idea of the overhead of each one? If the down stream is acking too slowly, I'd prefer that Vector start applying back-pressure to the source than run out of memory.
> 
> For overhead, each ack requires a `u64` (for the id) and a `oneshot::channel` (for communicating status). I'm not sure what a typical number of outstanding acks would be, but I see the value in having a cap. From the Splunk side, the default number of ack statuses they store per channel is 1_000_000 which implies that, by default, a client would at most query for 1_000_000 acks. Is that a reasonable cap? Should we let this be user-configurable?

👍 I think that'd be reasonable to cap at as a default, but also expose it as a configurable option per channel.

File: src/sinks/splunk_hec/common/service.rs
Comment (jszwedko) [reply src/sinks/splunk_hec/common/service.rs]: 👍 looks good, thank you!

File: src/sinks/splunk_hec/common/service.rs
Comment (tobz) [inline src/sinks/splunk_hec/common/service.rs:95]: So, this may currently work due to the number of wakeup sources present in `Driver`, but the code is at risk of entering a deadlocked state since the implicit contract of `poll_ready` is to ensure that when it returns `Poll::Pending`, something will wake-up the caller of `poll_ready` when there _is_ capacity.  As written, nothing is being told to wake up the caller when `self.max_pending_acks_reached` switches back to `false`.

A more robust strategy would be to use `tokio::sync::Semaphore`, wrapped with `tokio_util::sync::PollSemaphore`. A semaphore is essentially a fixed number of "permits" that can be acquired and returned.  If you try to acquire a permit and there are none available, you wait (and get woken up) when a permit becomes available.  In this case, you would construct the semaphore with a number of permits matching the "max pending acks" configuration value.

Thus, you would end up changing the service to try to acquire a permit in `poll_ready` via `PollSemaphore::poll_acquire`.  Since the permit needs to travel with the future generated by `call`, you'd have to store this internally in an `Option` or something, and then grab it in `call`.  Add the appropriate panics if `poll_ready` is called and there's already a permit stored, or if `call` is stored and there's no stored permit, etc.

Crucially, though, by calling `PollSemaphore::poll_acquire` in `poll_ready`, you would ensure that a wake-up chain was established so when there were no permits available, the tasdk would be woken up as soon as a permit was released back, when a future generated by `call` finally completed and was dropped.

**Review by tobz — CHANGES_REQUESTED** [review]

File: src/sinks/splunk_hec/common/service.rs
Comment (001wwang) [reply src/sinks/splunk_hec/common/service.rs]: Thanks for the detailed explanation! I've refactored this in a3517c6. One question re:
> Add the appropriate panics if poll_ready is called and there's already a permit stored

The rustdoc describing `poll_ready` states that 

>  Once `poll_ready` returns `Poll::Ready(Ok(()))`, a request may be dispatched to the service using `call`. Until a request is dispatched, repeated calls to `poll_ready` must return either `Poll::Ready(Ok(()))` or `Poll::Ready(Err(_))`.

so, I decided to return `Poll::Ready(Err(_))` rather than panic if a permit is already stored. That said, would it be valid to instead check whether a permit is already stored and, if so, simply keep returning `Poll::Ready(Ok(()))`? Given the docs, it seems like users of a Service might expect calling `poll_ready` multiple times to be valid (although I guess I'm not sure why you'd continue doing so after the initial `Poll::Ready(Ok(()))`).

File: src/sinks/splunk_hec/common/service.rs
Comment (tobz) [inline src/sinks/splunk_hec/common/service.rs:117]: There's a small bug here, I believe.

Since we first poll the semaphore for an available permit, we have to wait for the semaphore to give us a permit back before we actually check if we already have a permit.  If we assume that some downstream usage of our service is broken, and unintentionally calls `poll_ready` more than once, we _may_ deadlock due to the downstream caller waiting for us to tell them we're ready, while we wait for the semaphore to give us a permit that it may never be able to give us... because we're holding the last one.

It's slightly contrived, since the number for maximum pending acks is likely to be very big, but it's _technically_ possible.

I would probably rework the code so that we first match against `self.current_ack_slot.take()`, and if it's `Some(..)`, return the error, and otherwise, run through the logic you have after calling `poll_acquire`.  With that structure, you can also wrap the call to `poll_acquire` with `ready!` which will handle the `Poll::Pending` case for you.

File: src/sinks/splunk_hec/common/service.rs
Comment (tobz) [inline src/sinks/splunk_hec/common/service.rs:161]: Just a thought, non-blocking: I wonder if you could take advantage of the various optional values being used here to actually avoid returning `EventStatus::Delivered` at the end of every branch except for the rejected case, and boil it down to something like `ack_finalizer_tx.map(....).map(...).unwrap_or(EventStatus::Delivered)`.

**Review by tobz — CHANGES_REQUESTED** [review]

**Comment by tobz** [comment]
Since the file I was going to put the part two comment on isn't part of this PR, I gotta add it as an after-the-fact comment...

Anywho, part two of two regarding the `SplunkEventSent` event:

We currently _emit_ the actual `SplunkEventsSent` metric in the [encoder](https://github.com/vectordotdev/vector/blob/9ef6501e161e9ec003ea4cdda299e54051c8e224/src/sinks/splunk_hec/metrics/encoder.rs#L86), but per the [component specification](https://github.com/vectordotdev/vector/blob/master/docs/specs/component.md), we should be emitting it only after we've actually sent the event i.e. after the HTTP request gets a successful response.

File: src/internal_events/splunk_hec.rs
Comment (001wwang) [reply src/internal_events/splunk_hec.rs:23]: Happy to update the instrumentation. Would it be alright to follow-up on this (and other instrumentation feedback) in a separate PR though?

File: src/internal_events/splunk_hec.rs
Comment (001wwang) [reply src/internal_events/splunk_hec.rs:66]: Sure thing -- would like to follow up in a separate PR as mentioned above. I realize now the diff makes it look like these events are all newly added as part of this PR, but the non-ack related ones are legacy.

File: src/sinks/splunk_hec/common/service.rs
Comment (001wwang) [reply src/sinks/splunk_hec/common/service.rs:161]: Sure, will toy around with this -- think I ran into some issues last time I'd tried simplifying the many branches but forget exactly what they were.

File: src/sinks/splunk_hec/common/service.rs
Comment (001wwang) [reply src/sinks/splunk_hec/common/service.rs:117]: Ah I see, that makes sense. Will fix this. Just curious, re-referencing my question [here](https://github.com/vectordotdev/vector/pull/10135#discussion_r761567108), why are clients not supposed to call `poll_ready` after `Poll::Ready(Ok)`? I don't know why they would want to, but, in this case, I imagine we could check  if a permit already exists via `self.current_ack_slot.is_some()` and immediately return `Poll::Ready(Ok)` instead of `Err`.

File: src/sinks/splunk_hec/common/service.rs
Comment (tobz) [reply src/sinks/splunk_hec/common/service.rs]: Yeah, that's a fair point.  I've gone with panicking in the past because it usually indicates problems with the downstream implementation.

You can swap the `Err` for `Ok`, that's fine by me, so long as we don't throw away an already-acquired permit.

File: src/sinks/splunk_hec/common/service.rs
Comment (tobz) [reply src/sinks/splunk_hec/common/service.rs]: Yep, I'm fine with that. 👍🏻  (re: my response on the other question)

**Review by tobz — APPROVED** [review]

File: src/internal_events/splunk_hec.rs
Comment (jszwedko) [reply src/internal_events/splunk_hec.rs:45]: Yeah, we can remove it if we document the upgrade path in an upgrade guide for 0.19.0.

Comment (001wwang) [reply src/internal_events/splunk_hec.rs:45]: Cool, I'll include this change in the follow up instrumentation updates PR mentioned below.

**Review by jszwedko — APPROVED** [review]
I didn't do a thorough code review, but the added metrics and configuration options look good to me. We should still go back over the events to ensure they align with the component spec, but, as you noted in your comment, I'm happy to see this in a follow-up since these events were pre-existing. Thanks for making these updates!

**Pull request merged**
